### PR TITLE
BUG: use ellipkm1 in elliptical filter design to prevent numerical issues

### LINKED ISSUE #6384: BUG: Discrepancy in ellipord or ellip
As mentioned in https://github.com/scipy/scipy/pull/6328#issuecomment-231899630 there's some numerical error or something in `ellipord` or `ellip` (or maybe `sosfreqz`?)

These operations should produce a filter that is strictly below -150 dB at frequencies below 0.25, but it only reaches -147:

```
N, Wn = ellipord(0.3, 0.25, .5, 150)
sos = ellip(N, .5, 150, Wn, 'high', output='sos')
w, h = sosfreqz(sos)
dB = 20*np.log10(np.abs(h))
w /= np.pi
plot(w, dB)
```

shouldn't go over -150 db

### PR DESCRIPTION
**Pull request by roryyorke**
#### Reference issue

Closes gh-6384.

#### What does this implement/fix?
In computing the filter coefficients, calculations of the form ellipk(m) and ellipk(1-m) are used, with small m.  In this case the latter is more accurately computed with ellipkm1.

I also changed the inverse solvers for ellipk(m) / ellipk(1-m) and sn / cn when my initial ellipkm1 changes caused unit test failures.

The replaced `_kratio` function is a bit odd; it clips its argument to [0,1], but then uses abs(m); if `m > 1`, the return value is set to `-k_ratio`, but the *absolute* of the result is always returned.  I think the new implementation is simpler to understand.

The replaced `_v_ratio` and associated minimization worked with `abs(ineps - s /c)`.  `s/c` seems to be roughly analogous to `tan`, and like `tan` is notionally unbounded.  I've changed this to solve `inepsc * c - s`, which is bounded; the solution also takes advantage of the periodic nature of s and c to bracket the solution (`s/c` is, obviously, also periodic).  I observed in testing the previous solution would sometimes be in the second period, though this may not matter.

The 1000dB stop-band ripple unit test is not supposed to be practical, but is an extreme example requiring ellipkm1; the previously failing 150dB test also requires this fix, so the 1000dB test could be removed.

#### Additional information

I'm not knowledgeable on elliptic filter design, and it seems to rely on relatively advanced mathematics; I hope there is someone with more expertise who can review what I've done. 

I also don't have access to Lutova, Tosic, and Evans, the reference mentioned in the code.  I have used [1] for guidance; other than Wikipedia, this seems to be the only source of information on elliptic filter design not behind a paywall.  This reference seems to use different conventions than the Scipy solution (I think it normalizes frequencies differently), and I found it generally difficult to follow.

The ellipap code seems sparsely commented to me, but perhaps it is just a faithful rendering of an algorithm from Lutova et al.

[1] https://www.matheonics.com/Tutorials/Elliptic.html

### CODE REVIEW

**Comment by rgommers** [comment]
Thanks @roryyorke!

@endolith would you want to review this PR?

File: scipy/signal/filter_design.py
Comment (endolith) [inline scipy/signal/filter_design.py:4236]: This should be at the top of the file

File: scipy/signal/filter_design.py
Comment (endolith) [inline scipy/signal/filter_design.py:4228]: I don't have an opinion on whether this change is good or not

**Review by endolith — CHANGES_REQUESTED** [review]
Most of it looks fine. I don't really understand the `abs(ineps - s /c)` part, though.  I could probably spend more time and figure it out, but here are some small changes to make first.

**Comment by roryyorke** [comment]
Thanks for the review! I'll only be able to look at this on Saturday.

File: scipy/signal/filter_design.py
Comment (roryyorke) [reply scipy/signal/filter_design.py:4327]: The version of `ellipap` in this PR fails for rp < 1e-15 (or about there; 1e-16 fails) even if I remove the `rp == 0` check.

For `N==1` this is from a divide-by-zero in the earlier `1 / (10 ** (0.1 * rp) - 1)`; this might be solvable with `expm1`.

For `N==2`, `rp==1e-16` results in `krat == 0`; that is disallowed by `_solve_kratio`.  Now that I look at it again, `_solve_kratio` should probably return 0 for `krat==0`.  However, besides this, `_solve_vratio` further down takes `1 / sqrt(10 ** (0.1 * rp) - 1)` as its argument; again; this is possibly again fixable by using `expm1`.

I can explore allowing small `rp` values via using `expm1` -- would this be useful?  I would have thought for a sufficiently small `rp` value one might as well use the equivalent Chebyshev design functions.

I could change the error message ("if rp is very small, consider using cheby2").

Is it naive to expect `ellip(2, 1e-7, 20, 0.1)` to give a similar result to `cheby2(2, 20, 0.1)` ? They're not at all close; for the version of ellipap of this PR:

```
root@e5a8651e374b:/home/scipy# python3.7
Python 3.7.5 (default, Nov  7 2019, 10:50:52) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from scipy.signal import cheby2, ellip
>>> cheby2(2, 20, 0.1, output='zpk')
(array([0.90445128-0.42657694j, 0.90445128+0.42657694j]), array([0.90449059-0.09550526j, 0.90449059+0.09550526j]), 0.09546596712030596)
>>> ellip(2, 1e-10, 20, 0.1, output='zpk')
(array([-0.9999231+0.0124011j, -0.9999231-0.0124011j]), array([-0.97362727-0.02832822j, -0.97362727+0.02832822j]), 0.9740392272955193)
```

Results for e4238172b, from which this PR is based, are similar:

```
root@e5a8651e374b:/home/scipy# python3.7
Python 3.7.5 (default, Nov  7 2019, 10:50:52) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from scipy.signal import cheby2, ellip
>>> cheby2(2, 20, 0.1, output='zpk')
(array([0.90445128-0.42657694j, 0.90445128+0.42657694j]), array([0.90449059-0.09550526j, 0.90449059+0.09550526j]), 0.09546596712030596)
>>> ellip(2, 1e-10, 20, 0.1, output='zpk')
(array([-0.99992397+0.01233108j, -0.99992397-0.01233108j]), array([-0.97361247-0.02831258j, -0.97361247+0.02831258j]), 0.9740239771319361)
```

File: scipy/signal/filter_design.py
Comment (roryyorke) [reply scipy/signal/filter_design.py:4228]: This allows _solve_vratio to handle large ineps values (so, small eps values, and small rp values).  Using `ineps - s / c`:

```
>>> _solve_vratio(1e10, 0.5)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/scipy/scipy/signal/filter_design.py", line 4250, in _solve_vratio
    maxiter=1000)
  File "/home/scipy/scipy/optimize/_root_scalar.py", line 246, in root_scalar
    r, sol = methodc(f, a, b, args=args, **kwargs)
  File "/home/scipy/scipy/optimize/zeros.py", line 776, in brentq
    r = _zeros._brentq(f, a, b, xtol, rtol, maxiter, args, full_output, disp)
ValueError: f(a) and f(b) must have different signs
```

Having said that, for this particular example the supposedly better variant (`c * ineps - s`) gives the same answer for `ineps` values of `1e9` and greater.

**Comment by roryyorke** [comment]
I see this now has merge conflicts; should I rebase on master, and force-push to this branch?

Also, should I squash the "MAINT" commit into the "BUG" one?

**Comment by endolith** [comment]
> should I rebase on master, and force-push to this branch?

Yes

> should I squash the "MAINT" commit into the "BUG" one?

If you want to.  I usually do things like that, squashing and rewriting history so that it's easier for others to follow the changes in each commit.

**Comment by roryyorke** [comment]
The latest commit give test failures on my machine, though I think it's unlikely to be related to the elliptical filter changes:

```
=================================== FAILURES ===================================
______________________________ TestNNLS.test_nnls ______________________________

self = <scipy.optimize.tests.test_nnls.TestNNLS object at 0x7f2f5703e450>

    def test_nnls(self):
        a = arange(25.0).reshape(-1,5)
        x = arange(5.0)
        y = dot(a,x)
>       x, res = nnls(a,y)
E       TypeError: _nnls.nnls() missing required argument 'n' (pos 3)

a          = array([[ 0.,  1.,  2.,  3.,  4.],
       [ 5.,  6.,  7.,  8.,  9.],
       [10., 11., 12., 13., 14.],
       [15., 16., 17., 18., 19.],
       [20., 21., 22., 23., 24.]])
self       = <scipy.optimize.tests.test_nnls.TestNNLS object at 0x7f2f5703e450>
x          = array([0., 1., 2., 3., 4.])
y          = array([ 30.,  80., 130., 180., 230.])

scipy/optimize/tests/test_nnls.py:21: TypeError
____________________________ TestNNLS.test_maxiter _____________________________

self = <scipy.optimize.tests.test_nnls.TestNNLS object at 0x7f2f56fc7450>

    def test_maxiter(self):
        # test that maxiter argument does stop iterations
        # NB: did not manage to find a test case where the default value
        # of maxiter is not sufficient, so use a too-small value
        rndm = np.random.RandomState(1234)
        a = rndm.uniform(size=(100, 100))
        b = rndm.uniform(size=100)
        with assert_raises(RuntimeError):
>           nnls(a, b, maxiter=1)
E           TypeError: _nnls.nnls() missing required argument 'n' (pos 3)

a          = array([[0.19151945, 0.62210877, 0.43772774, ..., 0.81920207, 0.05711564,
        0.66942174],
       [0.76711663, 0.70...29,
        0.49193928],
       [0.41890017, 0.09633171, 0.85793837, ..., 0.14655078, 0.70561967,
        0.84086351]])
b          = array([0.71668817, 0.57414704, 0.67041315, 0.7548132 , 0.28323194,
       0.53181497, 0.10010894, 0.93179996, 0.659156...94, 0.10829871, 0.09842516, 0.70277838, 0.41093927,
       0.64179768, 0.05716342, 0.34741402, 0.4595992 , 0.00274406])
rndm       = RandomState(MT19937) at 0x7F2F571687C0
self       = <scipy.optimize.tests.test_nnls.TestNNLS object at 0x7f2f56fc7450>

scipy/optimize/tests/test_nnls.py:33: TypeError
=========================== short test summary info ============================
FAILED scipy/optimize/tests/test_nnls.py::TestNNLS::test_nnls - TypeError: _n...
FAILED scipy/optimize/tests/test_nnls.py::TestNNLS::test_maxiter - TypeError:...
= 2 failed, 39993 passed, 2204 skipped, 102 xfailed, 6 xpassed in 1671.15s (0:27:51) =

   Normal return from subroutine COBYLA

   NFVALS =   50   F = 2.485185E+01    MAXCV = 1.999965E-10
   X = 4.955358E+00   6.666553E-01
```

**Comment by roryyorke** [comment]
I looked at using an `expm1`-based expression in place of the `10 ** x - 1` expressions in `ellipord` and `ellipap` (diff below).

This seems to work; see, e.g., [this notebook gist](https://gist.github.com/roryyorke/0d7243427d95dfbf2d57f1d31da7376c) showing improving match of elliptical to Chebyshev type-2 filters as rp decreases.  In this case (N=10, rs=20dB), the fit is obviously better going from `rp=1e-10` to `rp=1e-20`, and slightly better going to `rp=1e-30`.  So, there may be utility allowing tiny rp values for those interested in the pole and zero positions of the elliptical filter for these cases.

It's not obvious to me that this is useful for actual filters, though: `20 * np.log10(np.nextafter(1, -np.inf))` is -9.6e-14, so (I think?) ripple smaller than 1e-13 dB is more-or-less meaningless from a filtering point of view.  This is besides reasonable questions like "why not use a Chebyshev type-2 filter if rp < 0.01dB?".
 
It's not clear to me if this is worth pursuing further.  If it is, should all the `10 ** x - 1` expressions in `filter_design.py` be considered for similar treatment?  I missed at least one elliptical-filter related use in the diff below, in `band_stop_obj`, but these expressions appear in Butterworth and Chebyshev design code too.

```diff
diff --git a/scipy/signal/filter_design.py b/scipy/signal/filter_design.py
index 4cd09f9a8..24de39a6b 100644
--- a/scipy/signal/filter_design.py
+++ b/scipy/signal/filter_design.py
@@ -3976,6 +3976,13 @@ def cheb2ord(wp, ws, gpass, gstop, analog=False, fs=None):
     return ord, wn
 
 
+_POW10M1 = np.log(10)
+
+def _pow10m1(x):
+    """10 ** x - 1 for x near 0"""
+    return np.expm1(_POW10M1 * x)
+
+
 def ellipord(wp, ws, gpass, gstop, analog=False, fs=None):
     """Elliptic (Cauer) filter order selection.
 
@@ -4096,9 +4103,7 @@ def ellipord(wp, ws, gpass, gstop, analog=False, fs=None):
 
     nat = min(abs(nat))
 
-    GSTOP = 10 ** (0.1 * gstop)
-    GPASS = 10 ** (0.1 * gpass)
-    arg1_sq = (GPASS - 1.0) / (GSTOP - 1.0)
+    arg1_sq = _pow10m1(0.1 * gpass) / _pow10m1(0.1 * gstop)
     arg0 = 1.0 / nat
     d0 = special.ellipk(arg0 ** 2), special.ellipkm1(arg0 ** 2)
     d1 = special.ellipk(arg1_sq), special.ellipkm1(arg1_sq)
@@ -4265,8 +4270,7 @@ def _f_kratio(m, krat):
 
 def _solve_kratio(krat):
     """Solve for m in krat = ellipk(m) / ellipk(1-m)"""
-    # ellipkm1(0) is inf, so move one step up
-    lo = np.finfo(np.float64).tiny
+    lo = 0
     # ellipk(1) is inf, so move one step down
     hi = np.nextafter(1, -np.inf)
     klo = _f_kratio(lo, 0)
@@ -4314,14 +4318,15 @@ def ellipap(N, rp, rs):
         # Even order filters have DC gain of -rp dB
         return numpy.array([]), numpy.array([]), 10**(-rp/20)
     elif N == 1:
-        p = -sqrt(1.0 / (10 ** (0.1 * rp) - 1.0))
+        p = -sqrt(1.0 / _pow10m1(0.1 * rp))
         k = -p
         z = []
         return asarray(z), asarray(p), k
 
-    eps_sq = 10 ** (0.1 * rp) - 1
+    eps_sq = _pow10m1(0.1 * rp)
+
     eps = np.sqrt(eps_sq)
-    ck1_sq = eps_sq / (10 ** (0.1 * rs) - 1)
+    ck1_sq = eps_sq / _pow10m1(0.1 * rs)
     if ck1_sq == 0:
         raise ValueError("Cannot design a filter with given rp and rs"
                          " specifications.")
```

**Comment by tylerjereddy** [comment]
@roryyorke @endolith Can you summarize the status of this PR as we approach within 2 weeks of branching for the next release? Almost ready to merge or still matters to discuss?

**Comment by roryyorke** [comment]
@tylerjereddy: I believe I've fixed all the stylistic issues. 

The question of (very) small `rp` remains, and I don't know how important this is for users.  The existing code also disallows `rp==0`:
```
    eps = numpy.sqrt(10 ** (0.1 * rp) - 1)
    ck1 = eps / numpy.sqrt(10 ** (0.1 * rs) - 1)
    ck1p = numpy.sqrt(1 - ck1 * ck1)
    if ck1p == 1:
        raise ValueError("Cannot design a filter with given rp and rs"
                         " specifications.")
```
so there is at least is no *change* in that respect in this PR, and we could defer addressing `rp < 1e-15` to another issue.

I won't have time to look at this for a while, probably a few months at least.  It is a fairly small change, and could be deferred.

**Comment by rgommers** [comment]
This looks close to ready, but let's bump the milestone and wait for the green light from @endolith here.

**Comment by roryyorke** [comment]
I found a good reference for elliptical filter design at https://www.ece.rutgers.edu/~orfanidi/ece521/notes.pdf   .  I don't know if this has been peer-reviewed, but it was written by an academic and textbook author.  The text is well-written, and it's straightforward  to implement algorithms from the equations presented.  He also gives Matlab code: I glanced at this, but didn't use it as a basis for the new code.

The latest change replaces the two `root_scalar` calls with more specific solvers, taken from the reference above.  I've compared the approaches (I'll post detail later) and the specific solvers are almost always better.  With this code I can synthesize a 20th order filter with passband ripple of 3db and stop-band ripple of 3000dB; I can also synthesize a 10th order filter with a pass-band ripple of 1e-30 dB that is nigh-indistinguishable from a Chebyshev type-2 filter (again, will post detail later).  These two examples demonstrate how far the new code can be pushed; I doubt they're of practical use.

**Comment by roryyorke** [comment]
Follow-up part 1: solving the degree equation (`_solve_kratio` previously).

Here we find elliptical modulus `m` (associated with selectivity) from modulus `m1` (associated with discrimination) and filter order `n`.  The equation is `n * K(1-m)/K(m) = K(1-m1)/K(m1)`, where `K` is the complete elliptic integral.  The error in the figures below is `abs(n * K(1-m)/K(m) = K(1-m1)/K(m1))`, with an artificial minimum of machine precision to make the plot clearer.

In the figures I've shown `m` and the error just discussed for three cases: the proposed new solution ("nome") ([1], Eq (49)), a solution based on the Jacobian elliptical function `sn` ([1] Eq, (47)), and the previous root-finding solution ("num").

I've shown two cases, n=2 and n=10.  For n=10 I've shown two different `m1` intervals.

For n=2 and m1 < 1e-1, the nome solution has the least error.  For n=10, the nome solution has the least error for m1 < 1e-3, and it is not too much worse than the sn solution for 1e-3 < m1 < 1e-1.  For m1 > 1e-1, the sn solution has the least error, though the error is close to 1 in all cases for large m1.  `m1 = (10**(rp/10) - 1) / (10**(rs/20)-1)` is usually fairly small; e.g., for rp=3dB and rs=20dB, m1 = 0.01 (approx.), and m1 will get smaller with smaller rp and larger rs; the error for m1 > 1e-1 is thus not too concerning.

deg_eq_n2

deg_eq_n10

deg_eq_n10_largem

**Comment by roryyorke** [comment]
Follow-up part 2: solving the inverse Jacobian elliptic function `sc` (previous `_solve_vratio`) with Landen transformations.

I want to change the implementation of the inverse function (`_arc_jac_sn`) to error out after, say, 10 iterations (Orfanidis thinks 5 is enough for "practical" filters). 

See below for comparisons of the previous numerical `root_scalar` solution with the new Landen transformation-based solution; the Landen solution is better in almost almost all cases.

`err` in this case is `sc(asc(w,m), m) - w` for different w values. `asc` is the inverse sc ("arc sc") function.

asc_m1e-10

asc_m5e-1

asc_m0p9999

**Comment by roryyorke** [comment]
Follow-up part 3: filter design demonstrations.

The first is a set of 20th order low-pass filters with passband ripple of 3dB, and stop-band ripple from 100dB to 3000dB:

demo_limits2

The second is a comparison of 10th order elliptical and Chebyshev type-2 filters with a stopband ripple of 20 dB.  The elliptical filter has a passband ripple of 1e-30 dB, and associated frequency chosen to math the Chebyshev type-2 response.

cf_ellip_cheby2_rp1em30

**Comment by tylerjereddy** [comment]
@roryyorke @endolith Is it possible to get another update on where you both think this PR stands now?

It looks like quite a bit of hard work has happened here and many of the review comments seem positive.

**Comment by roryyorke** [comment]
@tylerjereddy I believe it's done; I don't plan anything further.

The most recent 2 commits (b5f3794 and 294124c) haven't been reviewed.

**Review by larsoner — APPROVED** [review]
Given what I've read above and the test changes that uniformly make things better (fixing an `xfail` and adding a case that fails on `master` but passes here) and what I understand from the code, LGTM +1 for merge. I'll merge tomorrow unless someone objects or wants more time to review (@endolith ?)

**Comment by tylerjereddy** [comment]
Let's put it in thanks @larsoner @endolith et al.

**Pull request merged**
